cdap-ui/app/features/hydratorplusplus/controllers/detail/top-panel-ctrl.js

```diff
@@ -53,6 +58,81 @@ class HydratorDetailTopPanelController {
       this.setState();
     });
     HydratorPlusPlusDetailNonRunsStore.registerOnChangeListener(this.setScheduleStatus.bind(this));
+    this.macrosList = '';
+    this.runTimeWidgetConfig = {
+      'widget-type': 'keyvalue',
+      'widget-attributes': {
+        'showDelimiter': 'false'
+      }
+    };
+    this.fetchMacros();
+  }
+  fetchMacros() {
+    const macroReducer = (prev, curr) => {
+      return prev.concat(this.myHelpers.objectQuery(curr, 'spec', 'properties', 'macros', 'lookupProperties') || []);
+    };
+    const parseMacros = macrosSpec => {
+      return macrosSpec
+        .reduce(macroReducer, [])
+        .map(macro => macro + ':')
+        .join(',');
+    };
+    let {namespace, app} = this.HydratorPlusPlusDetailRunsStore.getParams();
+    let preferenceParams = {
+      namespace: this.$state.params.namespace,
+      appId: this.$state.params.pipelineId,
+      scope: this.$scope
+    };
+    const errorHandler = err => console.log('ERRRO', err);
+    return this.myPipelineApi
+      .fetchMacros({
+        namespace,
+        pipeline: app
+      })
+      .$promise
+      .then(
+        res => this.macrosList = parseMacros(res),
+        err => this.macroError = err
+      )
+      .then(
+        () => this.myPreferenceApi.getAppPreference(preferenceParams).$promise,
+        errorHandler
+      )
+      .then(
+        res => this.syncPreferencesStoreWithMacros(res),
+        errorHandler
+      );
+  }
+  syncPreferencesStoreWithMacros(appPreferences = {}) {
+    try {
+      appPreferences = JSON.parse(angular.toJson(appPreferences));
```

Comment:
why is this necessary? seems like we are turning something to JSON and parsing it back?